### PR TITLE
Concurrent hit misses due to Transients entry creation collision

### PR DESCRIPTION
**Pull request by rousskov**
Transients::addEntry() fails when two simultaneous cache hit requests
arrive at two different workers:

    Controller.cc find: e:=msV/0x55c2dbdac7f0*0 check failed: slot
    exception location: Transients.cc(236) addEntry

The failure leads to a cache miss for one of the requests. The miss
transaction will overwrite the already cached entry, possibly creating
more misses due to long write locks used by the disk cache. The problem
does not affect collapsed requests. It affects regular cache hits.

The failure happens because two requests are trying to create a
Transients entry at about the same time. Only one can succeed. With the
current ReadWriteLock-based design, some contention is unavoidable, but
the contention window was rather large/long:

* The critical section essentially started when Controller::find() (in
  each request worker) checked Transients and did not find a matching
  entry there. That failed lookup put each request into "create a new
  Transients entry" mode.

* A worker exited that critical section when openForWriting() in
  Transients::addEntry() returned (with a write lock to the
  ready-to-be-filled entry for one of the two requests and with a
  failure for the other).

This change reduces the contention window to the bare minimum necessary
to create and fill a Transients entry inside openOrCreateForReading().
In the micro-tests that exposed this problem, the probability of a
failure is reduced from more than 70% to less than 3%. YMMV.

Also split addEntry() into two directional methods because while they
are structured similarly they actually have almost no common code now.

### CODE REVIEW

File: src/ipc/StoreMap.cc [hunk 1]
```diff
@@ -99,6 +99,43 @@ Ipc::StoreMap::forgetWritingEntry(sfileno fileno)
...
+    // start with reading so that we do not overwrite an existing unlocked entry
+    auto idx = fileNoByKey(key);
+    if (const auto anchor = openForReadingAt(idx, key)) {
+        fileno = idx;
+        return anchor;
+    }
+
+    // the competing openOrCreateForReading() workers race to create a new entry
+    idx = fileNoByKey(key);
+    if (auto anchor = openForWritingAt(idx)) {
```
Comment (eduard-bagdasaryan) [inline src/ipc/StoreMap.cc:117]: Please consider moving this block into a new dedicated StoreMap::createForReadingAt(idx).

**Comment by eduard-bagdasaryan** [comment]
Also we need to s/Coordinator::find/Controller::find/ in the PR description.

**Comment by rousskov** [comment]
> Also we need to s/Coordinator::find/Controller::find/ in the PR description.

Fixed.

File: src/ipc/StoreMap.cc [hunk 2]
```diff
@@ -99,6 +99,43 @@ Ipc::StoreMap::forgetWritingEntry(sfileno fileno)
...
+{
+    debugs(54, 5, "opening/creating entry with key " << storeKeyText(key)
+           << " for reading " << path);
+
+    // start with reading so that we do not overwrite an existing unlocked entry
+    auto idx = fileNoByKey(key);
+    if (const auto anchor = openForReadingAt(idx, key)) {
+        fileno = idx;
+        return anchor;
+    }
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/ipc/StoreMap.cc:113]: Ok.

File: src/ipc/StoreMap.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/ipc/StoreMap.cc:117]: Ok.

File: src/ipc/StoreMap.cc [hunk 3]
```diff
@@ -99,6 +99,43 @@ Ipc::StoreMap::forgetWritingEntry(sfileno fileno)
...
+
+    // we lost the above race; see if the winner-created entry is now readable
+    // TODO: Do some useful housekeeping work here to give the winner more time.
+    idx = fileNoByKey(key);
+    if (const auto anchor = openForReadingAt(idx, key)) {
+        fileno = idx;
+        return anchor;
+    }
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/ipc/StoreMap.cc:133]: Ok.

**Comment by eduard-bagdasaryan** [comment]
Note that this PR has a problem with compilation:

```
Transients.cc:286:25: error: no match for ‘operator=’ (operand types are ‘MemObject::XitTable’ and ‘<brace-enclosed initializer list>’)
     e.mem_obj->xitTable = { index, Store::ioReading };
```

This happens because MemObject::XitTable::index member is default-initialized (e.g., see [here](https://stackoverflow.com/questions/18184096/c11-struct-initialization-compilation-error) for explanation).

**Comment by rousskov** [comment]
> This happens because MemObject::XitTable::index member is default-initialized

Thank you for finding that explanation! Fixed by 9e50eab. 

* I considered adding a xitTable constructor but it felt like an overkill, especially since C++14 removes the restriction that bit me.
* I decided to make all xitTable assignments consistent (and to reduce code duplication) instead of reducing the diff near Transients::addWriterEntry() by reusing the old addEntry() assignment there.

**Comment by rousskov** [comment]
@kinkie, when you get a chance, please check what seems like a Jenkins setup [problem](https://build.squid-cache.org/job/5-pr-auto/COMPILER=clang,OS=opensuse-leap,label=docker-build-host/742/console) affecting this PR:

    squid-farm-build: no such file or directory

**Comment by yadij** [comment]
@rousskov , @eduard-bagdasaryan  this PR looks to me like one of the ones important to have in v5. However each time I have attempted backport it has hit issues with code it depends on not being present in that branch.

There are some other minor stability fixes that in turn depend on this one. So if anyone is able to provide a working backport PR for v5 that would be a great help - otherwise this will have to wait in v6.

**Comment by rousskov** [comment]
Thank you for asking. We will backport this fix to v5 and post a PR. If that backport requires other backports, we will provide them as well. IIRC, all the related Store changes are appropriate for and desirable in v5.

**Pull request closed without merging**